neo/Ledger/MemoryPool.cs

```diff
@@ -4,162 +4,358 @@
+        }
+        
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool TryRemoveInternal(UInt256 hash, out Transaction tx)
+        {
+            if (_unsortedTransactions.TryRemove(hash, out PoolItem item))
             {
-                pool = _mem_pool_fee;
+                SortedSet<PoolItem> pool = item.Transaction.IsLowPriority
+                    ? _sortedLowPrioTransactions : _sortedHighPrioTransactions;
+                pool.Remove(item);
+                tx = item.Transaction;
+                return true;
             }
 
-            pool.TryAdd(hash, new PoolItem(tx));
+            tx = null;
+            return false;
+        }
 
-            if (Count > Capacity)
+        public void UpdatePoolForBlockPersisted(Block block, Snapshot snapshot)
+        {
+            // First remove the transactions verified in the block.
+            foreach (Transaction tx in block.Transactions)
             {
-                RemoveOldest(_mem_pool_free, DateTime.UtcNow.AddSeconds(-Blockchain.SecondsPerBlock * 20));
-
-                var exceed = Count - Capacity;
-
-                if (exceed > 0)
+                if (!TryRemoveInternal(tx.Hash, out _))
                 {
-                    RemoveLowestFee(_mem_pool_free, exceed);
-                    exceed = Count - Capacity;
-
-                    if (exceed > 0)
+                    if (_unverifiedTransactions.TryRemove(tx.Hash, out PoolItem item))
                     {
-                        RemoveLowestFee(_mem_pool_fee, exceed);
+                        SortedSet<PoolItem> pool = tx.IsLowPriority
+                            ? _unverifiedSortedLowPriorityTransactions : _unverifiedSortedHighPriorityTransactions;
+                        pool.Remove(item);
                     }
                 }
             }
 
-            return pool.ContainsKey(hash);
-        }
-
-        public bool TryRemove(UInt256 hash, out Transaction tx)
-        {
-            if (_mem_pool_free.TryRemove(hash, out PoolItem item))
+            // Add all the previously verified transactions back to the unverified transactions
+            foreach (PoolItem item in _sortedHighPrioTransactions)
             {
-                tx = item.Transaction;
-                return true;
+                _unverifiedTransactions.TryAdd(item.Transaction.Hash, item);
+                _unverifiedSortedHighPriorityTransactions.Add(item);
             }
-            else if (_mem_pool_free.TryRemove(hash, out item))
+            
+            // NOTE: This really shouldn't be necessary any more and can actually be counter-productive, since
+            //       they can be re-added anyway and would incur a potential addtional verification; setting to 1000
+            //       blocks for now. If transactions want to only be valid for short times they can include a script.
+            var lowPriorityCutOffTime = DateTime.UtcNow.AddSeconds(-Blockchain.SecondsPerBlock * 1000);
```

Comment:
In this sense, maybe this cutoff can be removed.

Reply:
We could remove it, but I guess if there were consistently low priority transactions transactions with small fees for a long period of time (like say 24 hours) it would be desirable to eventually expire off the completely free transactions with no fees to allow them to be sent again with a fee. Using 1000 blocks here achieves that; might as well keep it here; though clients should still use a best practice of attaching a script to expire transactions sooner; this is more like a maximum for them to use. Of course it is possible they could be expired sooner if the pool is filled up with 50000 higher fee transactions.